packages/detectors.py

```diff
@@ -0,0 +1,256 @@
+
+    """
+    The second part of the code determines which voxels are inside the brain
+    and which are outside the brain and creates a mask (boolean matrix)
+    """
+
+    #get the mean voxel intensity of entire 4D object
+    mean_voxel = np.mean(data)
+    #get the mean volume (3D) across time series (axis 3)
+    mean_volume = np.mean(data, axis=3)
+    #boolean mask set to all voxels above .5 in the first volume
+    #(.125 is the SPM criterion but .5 seems like a better threshold)
+    mask = mean_volume > (.5 * mean_voxel) #threshold can be adjusted!
+    out_mask = ~mask
+
+    """
+    The third part of code finds the root mean square of U from step 1, then uses the
+    mask from step 2 to determine which components explain data outside the brain
+    Selects these "bad components" with high "outsideness"
+    """
+
+    #Apply mask to C matrix to get all voxels outside of brain
+    outside = C_vols[out_mask]
+    #Get RMS of the voxels outside, reflecting "outsideness" of this scan
+    RMS_out = np.sqrt(np.mean((outside ** 2), axis=0))
+
+    #Apply mask to C matrix to get all voxels inside brain
+    inside = C_vols[mask]
+    #Get RMS of the voxels inside, reflecting "insideness" of this scan
+    RMS_in = np.sqrt(np.mean((inside ** 2), axis=0))
+
+    #The closer this ratio is to 1, the worse the volume
+    RMS_ratio = RMS_out / RMS_in
+
+    """
+    The fourth part of the code uses the "bad components" to generate a new
+    "bad data set" and then puts this dataset through the outlier detector
+    """
+
+    #Create a boolean mask for the 10% worst PCs (meaning highest RMS ratio)
+    PC_bad = np.percentile(RMS_ratio, 90)
+    PC_bad_mask = RMS_ratio > PC_bad
+
+    U_bad = U[:, PC_bad_mask]
+    C_bad = C[PC_bad_mask]
+
+    #generates data set based on the bad PCs and (U and C matrices)
+    X_bad = U_bad.dot(C_bad).T.reshape((vol_shape + (n_vols,)))
+
+    # calculate outliers using iqr_detector
+    _, outliers = mah_detector(X_bad)
+
+    return X_bad, outliers
+
+def dvars_detector(data):
+    """
+    Calculate outliers based on RMS of differences between volumes
+    Each volume is subtracted from the following volume, then the RMS
+    is calculated for each volume difference
+
+    Outliers are then calculated by passing the RMS differences to the
+    iqr_detector function.
+
+    Parameters
+    ----------
+    data : 4D array
+        4D fMRI data
+
+    Returns
+    -------
+    dvars : 1D array
+        1d array of RMS differences of volumes equal in number to timepoints-1
+    outliers : 1D array
+        1d array of indices of outlying timepoints
+    """
+    # init dvars
+    dvars = []
+    # for 0 to timepoints-1
+    for i in range(data.shape[-1] - 1):
+        # get difference between volumes
+        vol_diff = data[...,i + 1] - data[..., i]
+        # calculate rms
+        dvars.append(np.sqrt(np.mean(vol_diff ** 2)))
+    # get outliers using iqr_detector
+    _, outliers = iqr_detector(dvars)
```

Comment:
Ditto here - suggest returning the dvars measure and then calling the `iqr_detector` function on the returned vector.